firebase-segmentation/src/main/java/com/google/firebase/segmentation/remote/SegmentationServiceClient.java

```diff
@@ -14,43 +14,182 @@
 
 package com.google.firebase.segmentation.remote;
 
+import androidx.annotation.NonNull;
 import com.google.android.gms.tasks.Task;
+import com.google.android.gms.tasks.TaskCompletionSource;
 import com.google.android.gms.tasks.Tasks;
+import com.squareup.okhttp.MediaType;
 import com.squareup.okhttp.OkHttpClient;
+import com.squareup.okhttp.Request;
+import com.squareup.okhttp.RequestBody;
+import com.squareup.okhttp.Response;
+import java.io.IOException;
 import java.util.concurrent.Executor;
 import java.util.concurrent.Executors;
+import org.json.JSONException;
+import org.json.JSONObject;
 
 /** Http client that sends request to Firebase Segmentation backend API. To be implemented */
 public class SegmentationServiceClient {
 
+  private static final String FIREBASE_SEGMENTATION_API_DOMAIN =
+      "firebasesegmentation.googleapis.com";
+  private static final String UPDATE_REQUEST_RESOURCE_NAME_FORMAT =
+      "projects/%s/installations/%s/customSegmentationData";
+  private static final String CLEAR_REQUEST_RESOURCE_NAME_FORMAT =
+      "projects/%s/installations/%s/customSegmentationData:clear";
+  private static final String FIREBASE_SEGMENTATION_API_VERSION = "v1alpha";
+
+  private static final MediaType JSON = MediaType.parse("application/json; charset=utf-8");
+
   private final OkHttpClient httpClient;
   private final Executor httpRequestExecutor;
 
   public enum Code {
     OK,
 
-    SERVER_INTERNAL_ERROR,
+    HTTP_CLIENT_ERROR,
+
+    CONFLICT,
+
+    NETWORK_ERROR,
 
-    ALREADY_EXISTS,
+    SERVER_ERROR,
 
-    PERMISSION_DENIED
+    UNAUTHORIZED,
   }
 
   public SegmentationServiceClient() {
     httpClient = new OkHttpClient();
     httpRequestExecutor = Executors.newFixedThreadPool(4);
```

Comment:
A meta-point about the implementation so far:

The SDK is already using 6 threads in total, even though all it does is it writes to shared prefs and makes http calls.

To make it more efficient consider doing the following:
1. Expose an async public 3p API
2. Have a **synchronous** internal implementation
3. Have public API (FirebaseSegmentation class) have one thread-pool that it uses to call into the synchronous implementation.

As an additional side-effect of doing it, the come will be come much more simple as it won't have Task chains in it, but rather have straightforward sequential code.

Reply:
Done, very good suggestion.